File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.cpp

class Solution {
public:
    
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int n=(nums1.size()+nums2.size())/2,a1=0,b1=0;
        nums1.push_back(INT_MAX);
        nums2.push_back(INT_MAX);
        if(nums1.size()+nums2.size()&1){
            while((a1+b1)!=n){
                if(nums1[a1]>nums2[b1]) b1++;
                else a1++;
            }
            return nums1[a1]>nums2[b1]? nums2[b1] : nums1[a1];
        }
        while((a1+b1)!=n-1){
            if(nums1[a1]>nums2[b1]) b1++;
            else a1++;
        }
        return ((nums1[a1]>nums2[b1]? nums2[b1++] : nums1[a1++]) + (nums1[a1]>nums2[b1]? nums2[b1] : nums1[a1]))/2.0;
    }
};

File: 0004-median-of-two-sorted-arrays/0004-median-of-two-sorted-arrays.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int m = nums1.length, n = nums2.length, left = 0, right = m;
        if (m > n) return findMedianSortedArrays(nums2, nums1);
        
        while (left <= right) {
            int lp = (left + right)/2, rp = (m+n+1)/2 - lp;
            int maxleftA = (lp == 0) ? Integer.MIN_VALUE : nums1[lp - 1];
            int minrightA = (lp == m) ? Integer.MAX_VALUE : nums1[lp];
            int maxleftB = (rp == 0) ? Integer.MIN_VALUE : nums2[rp-1];
            int minleftB = (rp == n) ? Integer.MAX_VALUE : nums2[rp];
            
            if (maxleftA > minleftB) right = lp - 1;
            else if (maxleftB > minrightA) left = lp + 1;
            else {
                if (((m+n)&1) == 1) return Math.max(maxleftA, maxleftB);
                return (Math.min(minrightA, minleftB) + Math.max(maxleftA, maxleftB))/2.0;
            }
        }
        return -1;
    }
}